code/layout-analyzer.js

```diff
@@ -130,22 +199,97 @@ window.addEventListener('DOMContentLoaded', () => {
-      sfuCount[finger] = 0;
-      skuCount[finger] = 0;
-    });
+  // Returns a custom iterator, similar to Rust’s std::slice::Windows.
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_generators
+  // https://doc.rust-lang.org/std/primitive.slice.html#method.windows
+  const arrayWindows = (arr, len) => ({
+    *[Symbol.iterator]() {
+      for (let i = 0; i <= arr.length - len; i++) {
+        yield arr.slice(i, i + len);
+      }
+    }
+  });
+
+  const computeNGrams = () => {
+    const ngrams = Object.fromEntries([
+        // Digrams
+        'sfb',  // Same Finger Bigram
+        'skb',  // Same Key Bigram
+        'lsb',  // Lateral Strech Bigram
+        'handChange',  // Two keys typed by different hands
+        'scisor',  // Roll with uncomfortable height difference between the keys
+        'extendedScisor',  // scisor + lsb
+        'inwardRoll',   // Roll in the pinky -> index direction
+        'outwardRoll',  // Roll in the index -> pinky direction
+
+        // Trigrams
+        'redirect',  // Two rolls going in different directions
+        'badRedirect', // Redirect that doesn’t use the index
+        'sfs',  // Same Finger Skipgram (sfb with other key in the middle)
+        'sks',  // Same Key Skipgram (skb with other key in the middle)
+        'other',  // unused, is just two simplo digrams, nothing to note.
+      ].map(digramType => [digramType, {}])
+    );
+
+    const buildNgramDict = (dict, ngramLength) => {
+      let total = 0;
+      const rv = {};
+
+      for (const [ngram, frequency] of Object.entries(dict)) {
+        let nextPendingDeadKey = undefined;
+        const totalKeySequence = Array.from(ngram).flatMap(charToKeys);
+
+        for (const keySequence of arrayWindows(totalKeySequence, ngramLength)) {
+          total += frequency;
+          if (totalKeySequence.some(key => key == undefined)) continue;
+
+          let [pendingDeadKey, name] = keySequence.reduce(([pendingDeadKey, acc], { keyCode, level }) => {
+            let char = keyboard.layout.keyMap[keyCode][level];
+            if (pendingDeadKey)
+              char = keyboard.layout.deadKeys[pendingDeadKey][char];
+
+            return char.length == 1
+                ? [undefined, acc + char]
+                : [char, acc];
+          }, [nextPendingDeadKey, '']);
+
+          if (pendingDeadKey) {
+            name += pendingDeadKey;
+            pendingDeadKey = undefined;
+          }
+
+          // PrettyPrint the ODK
+          // TODO: add a system for generic deadKeys ?
+          name = name.replaceAll('**', '★');
+
+          // I wanted a "Zig-style block expression", syntax might be stupid
+          nextPendingDeadKey = (() => {
+            const { keyCode, level } = keySequence[0];
+            const firstCharInSequence = keyboard.layout.keyMap[keyCode][level]
+            if (firstCharInSequence.length == 1) return undefined;
+            return pendingDeadKey != undefined
```

Comment:
When comparing to `undefined`, `!===` is an absolute must-have.

Please make sure this doesn’t change the results, because I don’t trust JS for such comparisons.

Reply:
May I know why ? I couldn’t find weird conversions in the JS console (also yes, I checked and it doesn’t changes the results)